File: arrays_and_hashing/valid_sudoku.ts
//https://leetcode.com/problems/valid-sudoku/

const isValidSudoku = (board: string[][]): boolean => {
  for (let row = 0; row < board.length; row++) {
    const set = new Set()
    for (let col = 0; col < board[row].length; col++) {
      const current = board[row][col]
      if (current === '.') continue
      if (set.has(current)) return false
      set.add(current)
    }
  }

  for (let col = 0; col < board[0].length; col++) {
    const set = new Set()
    for (let row = 0; row < board.length; row++) {
      const current = board[row][col]
      if (current === '.') continue
      if (set.has(current)) return false
      set.add(current)
    }
  }

  for (let rowOffset = 0; rowOffset < board.length; rowOffset += 3) {
    for (let colOffset = 0; colOffset < board[0].length; colOffset += 3) {
      const set = new Set()
      for (let r = rowOffset; r < rowOffset + 3; r++) {
        for (let c = colOffset; c < colOffset + 3; c++) {
          const current = board[r][c]
          if (current === '.') continue
          if (set.has(current)) return false
          set.add(current)
        }
      }
    }

  }
  return true
};


const test1 = [
  ["5", "3", ".", ".", "7", ".", ".", ".", "."],
  ["6", ".", ".", "1", "9", "5", ".", ".", "."],
  [".", "9", "8", ".", ".", ".", ".", "6", "."],
  ["8", ".", ".", ".", "6", ".", ".", ".", "3"],
  ["4", ".", ".", "8", ".", "3", ".", ".", "1"],
  ["7", ".", ".", ".", "2", ".", ".", ".", "6"],
  [".", "6", ".", ".", ".", ".", "2", "8", "."],
  [".", ".", ".", "4", "1", "9", ".", ".", "5"],
  [".", ".", ".", ".", "8", ".", ".", "7", "9"]
]

const test2 = [
  ["8", "3", ".", ".", "7", ".", ".", ".", "."],
  ["6", ".", ".", "1", "9", "5", ".", ".", "."],
  [".", "9", "8", ".", ".", ".", ".", "6", "."],
  ["8", ".", ".", ".", "6", ".", ".", ".", "3"],
  ["4", ".", ".", "8", ".", "3", ".", ".", "1"],
  ["7", ".", ".", ".", "2", ".", ".", ".", "6"],
  [".", "6", ".", ".", ".", ".", "2", "8", "."],
  [".", ".", ".", "4", "1", "9", ".", ".", "5"],
  [".", ".", ".", ".", "8", ".", ".", "7", "9"]
]

console.log(isValidSudoku((test2)))

File: arrays_and_hashing/go/valid-sudoku.go
// https://leetcode.com/problems/valid-sudoku/
package main

import (
	"fmt"
)

func main() {
	// board := [][]byte{
	// 	{'5', '3', '.', '.', '7', '.', '.', '.', '.'},
	// 	{'6', '.', '.', '1', '9', '5', '.', '.', '.'},
	// 	{'.', '9', '8', '.', '.', '.', '.', '6', '.'},
	// 	{'8', '.', '.', '.', '6', '.', '.', '.', '3'},
	// 	{'4', '.', '.', '8', '.', '3', '.', '.', '1'},
	// 	{'7', '.', '.', '.', '2', '.', '.', '.', '6'},
	// 	{'.', '6', '.', '.', '.', '.', '2', '8', '.'},
	// 	{'.', '.', '.', '4', '1', '9', '.', '.', '5'},
	// 	{'.', '.', '.', '.', '8', '.', '.', '7', '9'},
	// }

	board2 := [][]byte{
		{'8', '3', '.', '.', '7', '.', '.', '.', '.'},
		{'6', '.', '.', '1', '9', '5', '.', '.', '.'},
		{'.', '9', '8', '.', '.', '.', '.', '6', '.'},
		{'8', '.', '.', '.', '6', '.', '.', '.', '3'},
		{'4', '.', '.', '8', '.', '3', '.', '.', '1'},
		{'7', '.', '.', '.', '2', '.', '.', '.', '6'},
		{'.', '6', '.', '.', '.', '.', '2', '8', '.'},
		{'.', '.', '.', '4', '1', '9', '.', '.', '5'},
		{'.', '.', '.', '.', '8', '.', '.', '7', '9'},
	}
	// fmt.Println(isValidSudoku(board))
	fmt.Println(isValidSudoku(board2))
}

func isValidSudoku(board [][]byte) bool {
	return checkRows(board) && checkCols(board) && checkBoxes(board)
}

func checkRows(board [][]byte) bool {
	for _, row := range board {

		lookup := make(map[byte]bool)
		for _, e := range row {
			if e == '.' {
				continue
			}
			_, ok := lookup[e]
			if ok {
				return false
			}
			lookup[e] = true
		}
	}
	return true
}

func checkCols(board [][]byte) bool {
	for i := range board[0] {
		lookup := make(map[byte]bool)
		for j := range board {
			current := board[j][i]
			if current == '.' {
				continue
			}
			_, ok := lookup[current]
			if ok {
				return false
			}
			lookup[current] = true
		}
	}
	return true
}

func checkBoxes(board [][]byte) bool {
	for i := 0; i <= 6; i += 3 {
		for j := 0; j <= 6; j += 3 {
			lookup := make(map[byte]bool)
			for r := i; r < i+3; r++ {
				for c := j; c < j+3; c++ {
					current := board[r][c]
					if current == '.' {
						continue
					}
					_, ok := lookup[current]
					if ok {
						return false
					}
					lookup[current] = true
				}
			}
		}
	}
	return true
}
